android/app/src/main/java/com/simplecityapps/shuttle/ui/screens/library/genres/FastScroller.kt

```diff
@@ -0,0 +1,354 @@
+import androidx.compose.foundation.lazy.LazyColumn
+import androidx.compose.foundation.lazy.LazyListState
+import androidx.compose.foundation.lazy.rememberLazyListState
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.derivedStateOf
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableFloatStateOf
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.layout.onGloballyPositioned
+import androidx.compose.ui.layout.positionInParent
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.platform.LocalInspectionMode
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.IntOffset
+import androidx.compose.ui.unit.IntSize
+import androidx.compose.ui.unit.dp
+import com.simplecityapps.shuttle.ui.theme.AppTheme
+import kotlin.math.roundToInt
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+@Composable
+fun FastScroller(
+    getPopupText: (index: Int) -> String?,
+    modifier: Modifier = Modifier,
+    state: LazyListState = rememberLazyListState(),
+    track: @Composable BoxScope.() -> Unit = {
+        DefaultTrack(modifier = Modifier.fillMaxHeight())
+    },
+    thumb: @Composable () -> Unit = {
+        DefaultThumb()
+    },
+    popup: @Composable (index: Int) -> Unit = { currentItemIndex ->
+        DefaultPopup(text = getPopupText(currentItemIndex))
+    }
+) {
+    val coroutineScope = rememberCoroutineScope()
+    val density = LocalDensity.current
+
+    // Drag-related state.
+    var isDragging by remember { mutableStateOf(false) }
+    var dragThumbOffsetPx by remember { mutableFloatStateOf(0f) }
+    var initialThumbOffset by remember { mutableFloatStateOf(0f) }
+    var cumulativeDrag by remember { mutableFloatStateOf(0f) }
+    var measuredThumbY by remember { mutableFloatStateOf(0f) }
+    var measuredThumbSize by remember { mutableStateOf(IntSize.Zero) }
+    var isVisible by remember { mutableStateOf(true) }
+
+    // Auto-hide the scroller when not scrolling or dragging.
+    LaunchedEffect(state.isScrollInProgress, isDragging) {
+        if (!state.isScrollInProgress && !isDragging) {
+            delay(1500)
+            if (!state.isScrollInProgress && !isDragging) {
```

Comment:
Is this second check redundant or is it needed because the state might've changed while on the `delay`?

Reply:
Yes it might have changed, so we need to check the condition again